plugin/pkg/admission/concurrentupdate/admission.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package concurrentupdate
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/golang/glog"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apiserver/pkg/admission"
+
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog"
+	scadmission "github.com/kubernetes-incubator/service-catalog/pkg/apiserver/admission"
+	informers "github.com/kubernetes-incubator/service-catalog/pkg/client/informers_generated/internalversion"
+	internalversion "github.com/kubernetes-incubator/service-catalog/pkg/client/listers_generated/servicecatalog/internalversion"
+)
+
+const (
+	// PluginName is name of admission plug-in
+	PluginName = "ConcurrentUpdate"
+)
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(io.Reader) (admission.Interface, error) {
+		return NewConcurrentUpdate(), nil
+	})
+}
+
+// concurrentUpdate is an implementation of admission.Interface.
+// It blocks concurrent updates to ServiceInstance and ServiceInstanceCredential
+// resources.
+type concurrentUpdate struct {
+	*admission.Handler
+	instanceLister           internalversion.ServiceInstanceLister
+	instanceCredentialLister internalversion.ServiceInstanceCredentialLister
+}
+
+var _ = scadmission.WantsInternalServiceCatalogInformerFactory(&concurrentUpdate{})
+
+func (cu *concurrentUpdate) Admit(a admission.Attributes) error {
+	// we need to wait for our caches to warm
+	if !cu.WaitForReady() {
+		return admission.NewForbidden(a, fmt.Errorf("not yet ready to handle request"))
+	}
+
+	if a.GetResource().Group != servicecatalog.GroupName {
+		return nil
+	}
+
+	// Do not block updates to the status of the resource as those updates need
+	// to succeed during reconciliation
+	if a.GetSubresource() == "status" {
+		return nil
+	}
+
+	if a.GetResource().GroupResource() == servicecatalog.Resource("serviceinstances") {
+		instance, ok := a.GetObject().(*servicecatalog.ServiceInstance)
+		if !ok {
+			return apierrors.NewBadRequest("Resource was marked with kind ServiceInstance but was unable to be converted")
+		}
+		if instance.DeletionTimestamp != nil || instance.Status.ReconciledGeneration != instance.Generation {
```

Comment:
I'm curious about whether we need the check on deletion timestamp - are you positive that we do?

Reply:
You are absolutely correct. I did not realize that the registry incremented the Generation of the resource on delete. So, the check for ReconciledGeneration != Generation covers deleting as well.